flocker/node/agents/blockdevice.py

```diff
@@ -527,23 +553,28 @@ class AttachVolume(PRecord):
     :ivar unicode hostname: An identifier for the node to which the volume
         should be attached.  An IPv4 address literal.
     """
-    volume = _volume_field()
+    dataset_id = field(type=UUID, mandatory=True)
     hostname = field(type=unicode, mandatory=True)
 
     @property
     def eliot_action(self):
-        return ATTACH_VOLUME(_logger, volume=self.volume)
+        return ATTACH_VOLUME(_logger, dataset_id=self.dataset_id)
 
     def run(self, deployer):
         """
         Use the deployer's ``IBlockDeviceAPI`` to attach the volume.
         """
-        # Make this asynchronous after FLOC-1549, probably as part of
-        # FLOC-1593.
-        deployer.block_device_api.attach_volume(
-            self.volume.blockdevice_id, self.hostname
+        api = deployer.async_block_device_api
+        listing = api.list_volumes()
+        listing.addCallback(
+            _blockdevice_volume_from_datasetid, self.dataset_id
         )
-        return succeed(None)
+
+        def found(volume):
+            ATTACH_VOLUME_DETAILS(volume=volume).write(_logger)
```

Comment:
What if _blockdevice_volume_from_datasetid returns `None` ? `dataset_id` not found among the volumes....for some reason? This log message will blow up with a type error or the attach_volume below will. Which will trigger an errback...which will be handled and logged by...?